storage_service::merge_topology_snapshot: handle big mutations

gusev-p opened the pull request:
The group0 state machine calls `merge_topology_snapshot` from `transfer_snapshot`. It feeds it with `raft_topology_snapshot` returned from `raft_pull_topology_snapshot`. This snapshot includes the entire `system.cdc_generations_v3` table. It can be huge and break the commitlog `max_record_size` limit.

The `system.cdc_generations_v3` is a single-partition table, so all the data is contained in one mutation object. To fit the commitlog limit we split this mutation into many smaller ones and apply them in separate `database::apply` calls. That means we give up the atomicity guarantee, but we actually don't need it for `system.cdc_generations_v3` and `system.topology_requests`.

This PR fixes the dtest `update_cluster_layout_tests.py::TestLargeScaleCluster::test_add_many_nodes_under_load`

fixes #17545

kbr-scylla commented:
> That means we give up the atomicity guarantee, but we actually don't need it for system.cdc_generations_v3 and system.topology_requests.

Should explain why.

I understand why we don't need atomicity for `system.cdc_generations_v3` -- because the data is not used by anything unless pointed to by entry inside `system.topology`. So it's fine to apply `system.cdc_generations_v3` nonatomically, followed by applying `system.topology` atomically.

But what's the argument for `system.topology_requests`?

denesb reviewed (approved):
LGTM (I mostly reviewed the mutation splitting part). 2 minor comments only.

gusev-p commented on service/storage_service.cc:
```diff
@@ -698,28 +699,123 @@ future<> storage_service::topology_transition() {
 }
 
+namespace {
+    future<> split_mutation(mutation source, std::vector<mutation>& target, size_t max_size) {
```
> Why not return `future<std::vector<mutation>>` instead of output param?

It's more 'dependency inject-y'/'composable'  - e.g. we can send output mutations from several `split_mutation` calls into the same output vector, like we do here.

gusev-p replied:
> Also, maybe this is worth putting somewhere accessible, so other code can use it if needed.

I would be glad to, but I haven’t found a suitable place. Separate `.hh/.cc`?

gusev-p replied:
I'll put it into `mutation.hh/.cc`.

avikivity replied:
I guess the best way would be to co_yield the mutations. This way it's clear it's an output parameter.

avikivity replied:
e.g. seastar::experimental::generator

gusev-p replied:
seems too fancy

gusev-p commented:
> Should explain why in commit messages.
> 
> I understand why we don't need atomicity for `system.cdc_generations_v3` -- because the data is not used by anything unless pointed to by entry inside `system.topology`. So it's fine to apply `system.cdc_generations_v3` nonatomically, followed by applying `system.topology` atomically.
> 
> But what's the argument for `system.topology_requests`?

I actually don't know/don't care. The original problem was with `system.cdc_generations_v3`, so I reverted the `system.topology_requests` part + added the arguments for `system.cdc_generations_v3`  to the commit message. The `system.topology_requests` shouldn't be big, as we see in [this log](https://github.com/scylladb/scylladb/issues/17545#issuecomment-1973414121).

gusev-p commented:
new version:
  * extracted `split_mutation` into `mutation.hh/mutation.cc`;
  * reverted the `system.topology_requests` part, we only care about `system.cdc_generations_v3`;
  * improved the commit message.

[the changes](https://github.com/scylladb/scylladb/compare/d31eee9351bbfc6ad4e7346805ea313101e7ad67..b062da75f97b7d7108d882bc367a8a71fd00de53).

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7146/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7146/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7146/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/285/)
#### Build Details:
- Duration: 4 hr 30 min
- Builder: spider8.cloudius-systems.com

avikivity commented:
There's no test for split_mutation.

Also, doesn't flat_mutation_reader do essentially the same?

avikivity commented:
Is there a safety issue? If we crash, only part of the mutation will be replayed.

Maybe we need the ability for commitlog to commit partial records, rather than hack around it in upper layers.

gusev-p commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7146/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7146/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7146/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/285/)
> 
> #### Build Details:
> * Duration: 4 hr 30 min
> * Builder: spider8.cloudius-systems.com

@yaronkaikov Why all the checks are green, but the CI still failed?

gusev-p commented:
> There's no test for split_mutation.

What exactly are you proposing to test here? There is no inner complexity in this function, it's just wiring existing components together, the efficiency ratio of a unit test would be low. The `test_cdc_generation_data.py` adjusted in this PR checks the function generally works.

> Also, doesn't flat_mutation_reader do essentially the same?

I don't understand this.

> Is there a safety issue? If we crash, only part of the mutation will be replayed.

This is mentioned in the commit message:
> That means we give up the atomicity guarantee, but we
> actually don't need it for system.cdc_generations_v3.
> The cdc_generations_v3 data is not used in any way until
> it's referenced from the topology table. By applying the
> cdc_generations_v3 mutations before topology mutations
> we ensure that the lack of atomicity isn't a problem here.

The data in `cdc_generations_v3` is not used until it's referenced from `topology`. For them to be referenced from `topology` we need to successfully apply `topology` mutations, and we make applying `cdc_generations_v3` mutations a prerequisite for that by applying them before `topology` mutations.

> Maybe we need the ability for commitlog to commit partial records, rather than hack around it in upper layers.

This sounds like a big effort, we strive to stay focused on fixing issues for 6.0.

denesb commented:
> * why not solve it closer to source? E.g. somewhere around system_keyspace::query_mutations where those huge mutations are generated. assuming that commitlog max mutations size will be typically the same across nodes (don't know if it's valid assumption - if not perhaps some cluster minimum could be maintained by raft)

I don't think `system_keyspace::query_mutations()` should make assumptions about what the mutations will be used for.

kbr-scylla commented:
> why the problem is limited to cdc_generations_v3 ? To me looks like potentially it can happen with everything which is passed via raft.

It can happen whenever you apply a mutation in any context

> is the problem limited to snapshot code only?

no, for example with cdc_generations, the topology coordinator code that creates and writes the generation mutations for the first time takes care to split them into multiple commands if necessary (see `prepare_and_broadcast_cdc_generation_data` in topology_coordinator.cc). The snapshot code needed to be adjusted as well which this PR does.

nuivall commented:
> > * why not solve it closer to source? E.g. somewhere around system_keyspace::query_mutations where those huge mutations are generated. assuming that commitlog max mutations size will be typically the same across nodes (don't know if it's valid assumption - if not perhaps some cluster minimum could be maintained by raft)
> 
> I don't think `system_keyspace::query_mutations()` should make assumptions about what the mutations will be used for.

It doesn't need to assume this implicitly but can have an option with max_size.

avikivity commented:
> > There's no test for split_mutation.
> 
> What exactly are you proposing to test here? There is no inner complexity in this function, it's just wiring existing components together, the efficiency ratio of a unit test would be low. The `test_cdc_generation_data.py` adjusted in this PR checks the function generally works.
> 
> > Also, doesn't flat_mutation_reader do essentially the same?
> 
> I don't understand this.

make_flat_mutation_reader_from_mutations() is basically a splitter.


> 
> > Is there a safety issue? If we crash, only part of the mutation will be replayed.
> 
> This is mentioned in the commit message:
> 
> > That means we give up the atomicity guarantee, but we
> > actually don't need it for system.cdc_generations_v3.
> > The cdc_generations_v3 data is not used in any way until
> > it's referenced from the topology table. By applying the
> > cdc_generations_v3 mutations before topology mutations
> > we ensure that the lack of atomicity isn't a problem here.
> 
> The data in `cdc_generations_v3` is not used until it's referenced from `topology`. For them to be referenced from `topology` we need to successfully apply `topology` mutations, and we make applying `cdc_generations_v3` mutations a prerequisite for that by applying them before `topology` mutations.
> 
> > Maybe we need the ability for commitlog to commit partial records, rather than hack around it in upper layers.
> 
> This sounds like a big effort, we strive to stay focused on fixing issues for 6.0.

It's too ad hoc. We sacrifice good engineering principles in the name of rushing out a release, then never fix it up.

kbr-scylla commented on mutation/mutation.cc:
```diff
@@ -194,6 +196,71 @@ mutation reverse(mutation mut) {
+        size_t _size = 0;
+        const size_t _max_size;
+        void maybe_flush() {
+            if (_size >= _max_size) {
```
Shouldn't we reset _size back to 0 after flushing?
I think this will cause us to start flushing after every row once we exceed _max_size for the first time.

kbr-scylla commented:
> There is no inner complexity in this function, it's just wiring existing components together, the efficiency ratio of a unit test would be low.

Hard disagree. The code is not trivial and I need to think to understand it. Tests would be valuable.

kbr-scylla commented on mutation/mutation.cc:
```diff
@@ -194,6 +196,71 @@ mutation reverse(mutation mut) {
+        stop_iteration consume(clustering_row&& cr) {
+            _size += cr.memory_usage(*_schema);
+            _mut_builder->consume(std::move(cr));
+            maybe_flush();
```
So we can open a partition, then flush, then consume partition end? How does that work?

kbr-scylla replied:
Yeah I know -- but how does that address my question?

In fact good that you referenced the `flush()` definition, which has a comment above it:
```cpp
    // Might only be called between consume_new_partition()
    // and consume_end_of_partition().
```
so my suspicions appear to be valid. The code IIUC is `flush()`ing in the middle of a partition, which is incorrect.

kbr-scylla replied:
@denesb maybe we should have checks in debug mode that verify such preconditions in `mutation_rebuilder{_v2}`

gusev-p replied:
I don't know what you are talking about. We _are_ between `consume_new_partition` and `consume_end_of_partition`.

kbr-scylla replied:
Ok (－‸ლ) I misunderstood the code and the comment. Indeed it makes sense that we flush in the middle of a partition and not in between partitions, that's the whole point

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -699,27 +699,58 @@ future<> storage_service::topology_transition() {
+        for (const auto& m: frozen_muts_to_apply) {
+            futures.push_back(_db.local().apply(s, m, {}, db::commitlog::force_sync::no, db::no_timeout));
+        }
+        co_await when_all_succeed(futures.begin(), futures.end()).discard_result();
```
Aren't you reimplementing `parallel_for_each`?
(Also perhaps we should limit concurrency with `max_concurrent_for_each`)

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -6179,6 +6179,24 @@ void storage_service::init_messaging_service(bool raft_topology_change_enabled)
                         return canonical_mutation{p.mut().unfreeze(s)};
                     });
+
+                    // Copy mutations n times, where n is picked so that the memory size of all mutations
```
I'd put the comment inside the injection so it's more clear when reading top-to-bottom that this is error injection specific logic.

avikivity commented on mutation/mutation.cc:
```diff
@@ -194,6 +196,71 @@ mutation reverse(mutation mut) {
     return *std::move(mut).consume(reverse_rebuilder, consume_in_reverse::yes).result;
 }
 
+future<> split_mutation(mutation source, std::vector<mutation>& target, size_t max_size) {
```
I'm sure we have >1 equivalents of split_mutation elsewhere.

gusev-p replied:
> I'm sure we have >1 equivalents of split_mutation elsewhere.

I couldn't find one with the required semantics (split by size of mutation in bytes). I asked @denesb and he proposed to create a new function.

gusev-p commented:
> > I don't think that `query_mutations` is actually 'closer to the source', the problem now is with commitlog `max_entry_size`, and not all mutations from `query_mutations` are going to be submitted to the commitlog.
> 
> see [#17632 (comment)](https://github.com/scylladb/scylladb/pull/17632#issuecomment-1980541154)

I still don't see why `query_mutations` should care about the size of mutations it produces and how it's better than a separate method.

nuivall commented:
> > > I don't think that `query_mutations` is actually 'closer to the source', the problem now is with commitlog `max_entry_size`, and not all mutations from `query_mutations` are going to be submitted to the commitlog.
> > 
> > 
> > see [#17632 (comment)](https://github.com/scylladb/scylladb/pull/17632#issuecomment-1980541154)
> 
> I still don't see why `query_mutations` should care about the size of mutations it produces and how it's better than a separate method.

I'm just trying to find some solution which would be a) relatively easy to implement b) would help fix other cases too.

Currently you modify only CDC.

(Also maybe it could spare some in memory copying).

As for a more theoretical reasoning: why first create too big mutations and then "fix" them later rather than request the right size at the beginning.

avikivity commented:
I'm sure we'll have the same problem with system.tablets in large clusters. We need a systematic fix.

denesb commented:
> As for a more theoretical reasoning: why first create too big mutations and then "fix" them later rather than request the right size at the beginning.

A generic producer of data like `query_mutations()` should not adapt to the particularities of any single end consumer. If we go down this route, producers like this start to accumulate all sorts of flags to tweak the output to the taste of different consumers. A much cleaner route is to add transforming stages to the pipeline, as Petr did in this PR. Or fix the end consumer (commitlog) to not be so picky (accept bigger mutations). This is a much better, generic solution, but it is more work.

gusev-p commented:
> make_flat_mutation_reader_from_mutations() is basically a splitter.

It produces `mutation_fragment`-s. The new function adds a bytes limit to it, and it reconstructs output mutations, since we need mutation-in-mutations-out interface.

gusev-p commented:
new version:
  * added test `random_mutation_generator`-based test for `split_mutation`;
  * use `max_concurrent_for_each`;
  * some small fixes in `split_mutation`.

@kbr-scylla

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7521/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7521/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8589/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7521/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/560/)
#### Build Details:
- Duration: 3 hr 23 min
- Builder: i-0c9ed4f7902f7c895 (m5d.12xlarge)

kbr-scylla commented on mutation/mutation.cc:
```diff
@@ -194,6 +196,89 @@ mutation reverse(mutation mut) {
+            // We could end up with an empty mutation if we consumed a range_tombstone_change
+            // and the next fragment exceeds the limit. The tombstone range may not have been
+            // closed yet and range_tombstone will not be created.
+            // This should be a rare case though, so just pop such mutation.
```
What happens when we flush mutation in between range tombstone changes?

Won't we incorrectly "forget" the range tombstone i.e. it won't end up in any of the mutations we produce?

kbr-scylla commented on mutation/mutation.cc:
```diff
@@ -194,6 +196,89 @@ mutation reverse(mutation mut) {
+    }
+    stop_iteration consume_end_of_partition() {
+        _state->builder.consume_end_of_partition();
+        if (auto mut_opt = _state->builder.consume_end_of_stream()) {
```
Could we assert that `mut_opt` is always engaged? Since we just consumed end of partition --- logically a mutation should be present.

gusev-p replied:
done

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -721,27 +721,55 @@ future<> storage_service::topology_transition() {
+        // By applying the cdc_generations_v3 mutations before topology mutations
+        // we ensure that the lack of atomicity isn't a problem here.
+        co_await max_concurrent_for_each(frozen_muts_to_apply, 128, [&] (const frozen_mutation& m) -> future<> {
+            return _db.local().apply(s, m, {}, db::commitlog::force_sync::no, db::no_timeout);
```
Although it probably doesn't matter what `force_sync` you pass here (since the table uses schema commitlog), but using `force_sync::no` is confusing. (And we do need to sync.)

gusev-p replied:
done

gusev-p commented:
new version:
  * use `force_sync::yes`;
  * add assert in `consume_end_of_partition`;

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7572/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7572/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/580/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8610/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7572/)
#### Build Details:
- Duration: 2 hr 5 min
- Builder: spider8.cloudius-systems.com

denesb commented on mutation/mutation.cc:
```diff
@@ -194,6 +196,89 @@ mutation reverse(mutation mut) {
+    stop_iteration consume_end_of_partition() {
+        _state->builder.consume_end_of_partition();
+        auto mut_opt = _state->builder.consume_end_of_stream();
+        assert(mut_opt);
```
Please avoid using asserts, they can bring down entire clusters. Use `on_internal_error()` instead.

gusev-p replied:
> they can bring down entire clusters

can you elaborate?

denesb replied:
We compile `assert()` into release builds and an `assert()` triggering will cause ScyllaDB to crash.
It is bad enough if this happens on a single node, but there has been cases where multiple nodes or entire clusters were brought down by `asserts()` triggering.
So asserts should be avoided as much as possible, and always prefer to use `on_internal_error()` instead.
Use `assert()` only to check invariants, the violation of which is unrecoverable and the node is not able to continue working correctly.

kbr-scylla replied:
but in this case there are corresponding `assert`s inside `builder`, and we can quickly infer that if those assertions pass then this assertion has to pass too.
(Need to double-check)

kbr-scylla replied:
Double checked, actually there's no such assert. (We're using `mutation_rebuilder_v2`, not `mutation_rebuilder` which has it)

The logic is correct I think, we only call `end_of_partition()` when we have an opened partition.

But I agree  `on_internal_error` would be better.

denesb reviewed (approved):
Still LGTM, one assert to be exorcised.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7607/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7607/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8638/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/606/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7607/)
#### Failed Tests (2/29798):
- [test_raft_recovery_basic](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7607/testReport/junit/%28root%29/test_raft_recovery_basic/Tests___Unit_Tests___test_raft_recovery_basic) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_raft_recovery_basic)
- [topology_custom.test_raft_recovery_basic.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7607/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_custom_test_raft_recovery_basic_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_raft_recovery_basic.debug.3)
#### Build Details:
- Duration: 2 hr 8 min
- Builder: spider1.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7611/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7611/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/609/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8641/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7611/)
#### Build Details:
- Duration: 2 hr 9 min
- Builder: spider3.cloudius-systems.com
